api/src/main/java/org/apache/iceberg/SnapshotUpdateValidator.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.function.BiConsumer;
+
+/**
+ * Interface to support validating snapshots during the commit process.
+ *
+ * <p>Validation will be called after the table metadata is refreshed but before the updates are
+ * committed to the table.
+ */
+@FunctionalInterface
+public interface SnapshotUpdateValidator
+    extends BiConsumer<Iterable<Snapshot>, Iterable<Snapshot>> {
+
+  SnapshotUpdateValidator NON_VALIDATING = (baseSnapshots, updatedSnapshots) -> {};
+
+  /**
+   * Validate the snapshots based on the refreshed and updated table state.
+   *
+   * <p>Implementations should throw a {@link org.apache.iceberg.exceptions.CommitFailedException}
```

Comment:
Shouldn't this be `ValidationException`? Throwing a `CommitFailedException` would trigger a retry or signal that the update is out of retries.

Reply:
Per above, you may want to trigger retry since you will do a reload, but you may also want to prevent retries.